### ISSUE TITLE
Not enough stack space to initialize exception classes to raise stack overflow

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#1743]
```ruby
def fib(n)
  if n <= 1
    n
  else
    fib(n - 1) + fib(n - 2)
  end
end

fib(1024)
```

You need some quite specific flags to trigger this!

```
$ graalvm-ce-19.2.0/Contents/Home/bin/ruby --jvm --vm.Dgraal.TruffleCompileOnly=fib --vm.Dgraal.Dump=Truffle:1 stack-overflow.rb 
org.graalvm.polyglot.PolyglotException: java.lang.NoClassDefFoundError: Could not initialize class org.truffleruby.language.control.RaiseException
	at org.truffleruby.language.methods.ExceptionTranslatingNode.translate(ExceptionTranslatingNode.java:94)
	at org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:57)
	at org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
	at org.truffleruby.language.methods.CatchForLambdaNode.execute(CatchForLambdaNode.java:48)
	at org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
	at org.graalvm.polyglot.Value.execute(Value.java:366)
	at org.truffleruby.launcher.RubyLauncher.runRubyMain(RubyLauncher.java:228)
	at org.truffleruby.launcher.RubyLauncher.launch(RubyLauncher.java:121)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:121)
	at org.graalvm.launcher.AbstractLanguageLauncher.launch(AbstractLanguageLauncher.java:70)
	at org.truffleruby.launcher.RubyLauncher.main(RubyLauncher.java:38)
Original Internal Error: 
java.lang.NoClassDefFoundError: Could not initialize class org.truffleruby.language.control.RaiseException
	at org.truffleruby.language.methods.ExceptionTranslatingNode.translate(ExceptionTranslatingNode.java:94)
	at org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:57)
	at org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
	at org.truffleruby.language.methods.CatchForLambdaNode.execute(CatchForLambdaNode.java:48)
	at org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
```

Works as intended with `--native`, with `--vm.Dgraal.Dump=`, and if you remove `--vm.Dgraal.TruffleCompileOnly=fib`.

**Comment by aardvark179** [comment]
I agree that's an annoying edge case, but it may be quite hard to completely solve. Throwing an exception will need some stack space for the creation of the exception itself, and may additionally need space to deoptimise methods as the stack is unwound, and both of these can result in the stack overflow exception itself causing a stack overflow.

You can avoid increase the stack size using an argument such as `--vm.Xss4m`, and once your fib function has compiled it should require dramatically less stack space than the interpreted version, but unless you disable background compilation I would expect your fib function to have reached its maximum stack depth well before the compilation has been completed.

**Comment by chrisseaton** [comment]
The JVM maintains several layers of reserve stack space so it can always throw itself.

https://pangin.pro/posts/stack-overflow-handling

I wonder if we can 'lift' that same idea to the guest-level, in the same way we did safepoints. Or otherwise configure these pages to allow us to do what we want?

**Comment by aardvark179** [comment]
So there are two things we'd need to consider, the stack spaces required to construct our exception object, and the stack space required to de-optimise methods as we unwidn the stack. The JVM itself can't guaranteee the latter, so has special code to ensure that monitors are released as the stack is unwound even if catch blocks can't actually be executed, and I think some special code round privileged operations.

We could maybe build a mechanism by which we always have a stack overflow exception ready, or we signal a different thread to create the exception to avoid adding to our own stack space, but would still rely on us being able to do that in less stack space than was consumed by the frames (interpreted or compiled) that generated the exception in the first place.

**Comment by eregon** [comment]
I think a simple workaround here would be to eagerly load `RaiseException` and other classes needed to handle the `StackOverflowError`.

BTW, we already print `[ruby] WARNING StackOverflowError` on STDERR just after catching the `StackOverflowError`.
The fact we don't see it here probably means `org.truffleruby.language.methods.TranslateExceptionNode#translate` fails to load `RaiseException`.

In any case, PR welcome.

**Comment by eregon** [comment]
Fixed in ebba39acffc174d525b0bd7f02dc74e9e2bbb73f
